### PR TITLE
Improve download zip

### PR DESCRIPTION
**Pull request by akaszynski**
Work on this was inspired by some issues with `_retreive_file`, particularly that when downloading and decompressing a zip, the contents of that zip file are placed directly into the downloads directory. This doesn't seem like an issue at first, but it can become one if two different archives were to contain identically named files that would be overwritten.

As you can imagine, this did involve a bit of a rework of our existing download internals:

- The existing `_retreive_file` function does double duty, downloading single files and downloading and unzipping zip archive files. This makes it a bit hard to follow and it's been split into `_retreive_file` and `_retreive_zip`.
- Various examples and download functions had to be rewritten to account for the various directory changes.
- To avoid having to manually clear out the cache, this PR adds in a `_verify_cache_integrity` function that's called when initializing `pyvista`. This ensures that caches that are out-of-date are automatically cleared.

### Bonus
Improve unit test stability by disabling download example on mac in https://github.com/pyvista/pyvista/pull/2727/commits/15e2488c05ba44c74455a67304b62f4c2abad7ad.

### CODE REVIEW

File: pyvista/examples/downloads.py [hunk 1]
```diff
@@ -3594,7 +3770,7 @@ def download_cavity(load=True):
 
     """
     directory, _ = _download_file('OpenFOAM.zip')
-    filename = os.path.join(directory, 'cavity', 'case.foam')
+    filename = os.path.join(directory, 'OpenFOAM', 'cavity', 'case.foam')
```
Comment (MatthewFlamm) [inline pyvista/examples/downloads.py:3773]: `directory` in the line above should be returning the path to the folder.  The `cavity/case.foam` are part of the internal structure of this zip file.  The resulting path I get with `examples.download_cavity(load=False)` is `'/path/to/pyvista/examples/OpenFOAM/OpenFOAM/cavity/case.foa
m'`, which shows that this top-level folder is duplicated.

Comment (MatthewFlamm) [reply pyvista/examples/downloads.py:3773]: This zip file, and it seems most of the others too, include a root level folder, presumably because of the very issue motivated by this PR.  One thought is to allow an optional parameter to put the base zip name in the path, or not, but it requires exposing that parameter in many functions.  Or we can leave as is, Im fine with that too.

Comment (akaszynski) [reply pyvista/examples/downloads.py:3773]: > This zip file, and it seems most of the others too, include a root level folder, presumably because of the very issue motivated by this PR. One thought is to allow an optional parameter to put the base zip name in the path, or not, but it requires exposing that parameter in many functions. Or we can leave as is, Im fine with that too.

Thats the intended consequence of this PR, and I’d prefer keeping the default behavior the only behavior.  We can choose to restructure the archive files (ideally with a new file), but unzipping the contents into a single directory might overwrite identically named files in the edge case that two archives contain identically names files.

File: pyvista/examples/downloads.py [hunk 2]
```diff
@@ -89,34 +120,114 @@ def _retrieve_file(retriever, filename):
...
+        HTTP download Response.
+
     """
     _check_examples_path()
     # First check if file has already been downloaded
     local_path = os.path.join(pyvista.EXAMPLES_PATH, os.path.basename(filename))
-    local_path_no_zip = local_path.replace('.zip', '')
-    if os.path.isfile(local_path_no_zip) or os.path.isdir(local_path_no_zip):
-        return local_path_no_zip, None
+    if os.path.isfile(local_path):
```
[the remark this answers is not in the document]
Comment (MatthewFlamm) [reply pyvista/examples/downloads.py:134]: The same applies for zip files that have the same name in different folders.

The only solution I can think of is to ensure that we get the same folder structure as `vtk-data/Data`

**Review by MatthewFlamm — COMMENTED** [review]
I don't see any changes to tests/plotting/test-*, but a change in a image cache file.  What is the reason?

File: pyvista/examples/downloads.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (MatthewFlamm) [reply pyvista/examples/downloads.py:134]: The same applies for zip files that have the same name in different folders.

The only solution I can think of is to ensure that we get the same folder structure as `vtk-data/Data`

**Comment by akaszynski** [comment]
> I don't see any changes to tests/plotting/test-*, but a change in a image cache file. What is the reason?

When the VRML example added a image cache test but forgot to add the image. I've moved this to #2765.

File: pyvista/examples/downloads.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (akaszynski) [reply pyvista/examples/downloads.py:135]: Implemented in ed791c5c6fc32f1148f60d049c440fe10075142b.

File: pyvista/examples/downloads.py [hunk 3]
```diff
@@ -130,11 +241,48 @@ def _download_file(filename):
...
+    -------
+    pyvista.DataSet or str
+        Dataset or path to the file depending on the ``load`` parameter.
+
+    """
     saved_file, _ = _download_file(filename)
+
+    # if zip contains a single file, permit read and load
+    if pyvista.get_ext(filename) == '.zip':  # pragma: no cover
+        files = os.listdir(saved_file)
```
[the remark this answers is not in the document]
Comment (akaszynski) [reply pyvista/examples/downloads.py]: > Does it make sense to even allow `_download_and_read` for zip files? I would vote for raising an Error here as it isn't clear what to do for a generic zip file.

Nope. Removed in 78b8b64be49bf1bea96964bbb23f212fca09d9bd.

File: pyvista/utilities/helpers.py [hunk 4]
```diff
@@ -1388,7 +1388,7 @@ def axis_rotation(points, angle, inplace=False, deg=True, axis='z'):
     return transformations.apply_transformation_to_points(rot_mat, points, inplace=inplace)
 
 
-def cubemap(path='', prefix='', ext='.jpg'):
+def cubemap(path='', prefix='', ext='.jpg', image_paths=None):
```
[the remark this answers is not in the document]
Comment (akaszynski) [reply pyvista/utilities/helpers.py:1391]: > These changes seem tangential to this PR, although maybe you had added a new zip file for additional testing?

This PR was originally going to just add this single feature (the `download_cube_map_debug` function), but after encountering the issues downloading files the scope changed.

Agree that it's tangential. I can move this, but the scope is at least somewhat related.

> I'm not a fan of having functions where 1 parameter usage completely invalidates all the other parameters. This suggests we should have two separate functions. In this case, it simplifies things further as one function can call the second one after assembling the correct list of filenames.

Agree that it's a back practice. Refactored in 5a512479f90bb24637ee7efe0a1e7deb18856f92.

**Comment by akaszynski** [comment]
@MatthewFlamm, please let me know if I've addressed your concerns in this PR.

**Review by MatthewFlamm — APPROVED** [review]
This looks good to me.  I think this is much better.  Thanks for implementing using the `vtk-data/Data` structure.  This gives me an additional idea which I'll submit an issue.
